core/src/main/java/org/infinispan/persistence/async/AsyncNonBlockingStore.java

```diff
@@ -469,36 +461,37 @@ public void run() {
    }
 
    CompletionStage<Void> submitModification(Modification modification) {
+      boolean isTraceEnabled = log.isTraceEnabled();
       boolean startNewBatch;
       CompletionStage<Void> submitStage;
       synchronized (this) {
+         int previousBatchId;
+         if (isTraceEnabled) {
+            previousBatchId = System.identityHashCode(replicatingModifications);
+            int currentBatchId = System.identityHashCode(pendingModifications);
+            log.tracef("Adding modification %s to batch %s", modification, currentBatchId);
+         } else {
+            previousBatchId = 0;
+         }
          modification.apply(this);
 
-         if (startNewBatch = batchFuture == null) {
+         startNewBatch = batchFuture == null;
+         if (startNewBatch) {
             batchFuture = new CompletableFuture<>();
          }
-         int replicatingSize = replicatingModifications.size();
 
-         submitStage = pendingModifications.size() + replicatingSize > modificationQueueSize
-               ? batchFuture : null;
-      }
-
-      boolean isTraceEnabled = log.isTraceEnabled();
-      if (isTraceEnabled) {
-         log.tracef("A new modification %s has been enqueued with async store", modification);
+         int queueSize = pendingModifications.size() + replicatingModifications.size();
+         submitStage = queueSize > modificationQueueSize ? batchFuture : null;
+         if (submitStage != null && isTraceEnabled) {
+            log.tracef("Too many modifications queued (%d), operation must wait until previous batch %d completes",
+                       queueSize, previousBatchId);
+         }
       }
 
       if (startNewBatch) {
-         if (log.isTraceEnabled()) {
-            log.tracef("Requesting a new async batch operation to be ran!");
-         }
          submitTask();
       }
-      if (submitStage != null && isTraceEnabled) {
-         log.tracef("Operation will not return immediately, must wait until current batch completes");
-      }
-      return submitStage == null ? CompletableFutures.completedNull() :
-            submitStage.thenApplyAsync(CompletableFutures.toNullFunction(), nonBlockingExecutor);
```

Comment:
Are you sure we can get rid of the apply async here? Without this the thread completing the submitStage will serially continue the processing for possibly hundreds of waiting requests.

Reply:
@wburns none of those waiting requests should be doing any blocking stuff, so I don't think it should be a problem.
My very simplistic thinking was that submitting to a different thread might add a performance cliff: the store can't keep up already, and we're adding thread contexts to make things even slower.

OTOH this change is a bit outside the declared scope of this PR, so I'll revert it.

Reply:
Done